### PR TITLE
Build node template in GCE cloud provider

### PR DESCRIPTION
**Pull request by mwielgus**
Step 2 for scale to/from 0.
Ref: #43
cc @MaciekPytel @fgrzadkowski

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go [hunk 1]
```diff
@@ -101,6 +107,17 @@ func (m *GceManager) RegisterMig(mig *Mig) {
...
+	template, err := m.getMigTemplate(mig)
+	if err != nil {
+		glog.Errorf("Failed to build template for %s", mig.Name)
+	}
+
+	node, err := m.buildNodeFromTemplate(mig, template)
+	if err != nil {
+		glog.Errorf("Failed to build template for %s", mig.Name)
+	}
+	glog.V(4).Infof("Node tempalte for mig %s - %#v", mig.Name, node)
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_manager.go]: s/tempalte/template

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go [hunk 2]
```diff
@@ -101,6 +107,17 @@ func (m *GceManager) RegisterMig(mig *Mig) {
 	m.migs = append(m.migs, &migInformation{
 		config: mig,
 	})
+
+	template, err := m.getMigTemplate(mig)
+	if err != nil {
+		glog.Errorf("Failed to build template for %s", mig.Name)
+	}
+
+	node, err := m.buildNodeFromTemplate(mig, template)
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_manager.go]: Are you building this here to save it and return from Mig.TemplateNodeInfo()? If so why not include that part in this PR?

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go [hunk 3]
```diff
@@ -248,3 +265,131 @@ func (m *GceManager) GetMigNodes(mig *Mig) ([]string, error) {
...
+	// TODO - setup status
+	return &node, nil
+}
+
+func buildGenericLabels(ref GceRef, machineType string, nodeName string) (map[string]string, error) {
+	result := make(map[string]string)
+
+	// TODO: extract it somehow
+	result[metav1.LabelArch] = "amd64"
+	result[metav1.LabelOS] = "linux"
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_manager.go]: nitpicking, but can we make those consts?

File: cluster-autoscaler/cloudprovider/gce/gce_manager_test.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	labels, err := buildGenericLabels(GceRef{
+		Name:    "kubernetes-minion-group",
+		Project: "mwielgus-proj",
+		Zone:    "us-central1-b"},
+		"n1-standard-8", "sillyname")
+	assert.Nil(t, err)
+	assert.Equal(t, "us-central1", labels[metav1.LabelZoneRegion])
+	assert.Equal(t, "us-central1-b", labels[metav1.LabelZoneFailureDomain])
+	assert.Equal(t, "sillyname", labels[metav1.LabelHostname])
+	assert.Equal(t, "n1-standard-8", labels[metav1.LabelInstanceType])
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_manager_test.go]: I know it's trivial but can we also assert os / arch labels?

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go [hunk 5]
```diff
@@ -248,3 +265,131 @@ func (m *GceManager) GetMigNodes(mig *Mig) ([]string, error) {
...
+
+	// TODO: handle custom !!!!
+	// TODO: handle GPU
+	machineType, err := m.service.MachineTypes.Get(mig.Project, mig.Zone, template.Properties.MachineType).Do()
+	if err != nil {
+		return nil, err
+	}
+	node.Status.Capacity[apiv1.ResourceCPU] = *resource.NewQuantity(machineType.GuestCpus, resource.DecimalSI)
+	node.Status.Capacity[apiv1.ResourceMemory] = *resource.NewQuantity(machineType.MemoryMb, resource.DecimalSI)
+	node.Status.Allocatable = node.Status.Capacity
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_manager.go]: That is not true in the clusters I run on GCE with kube-up.sh (though it's pretty close). We should at least add a TODO, until we figure it out. Which I think is quite important, as I can imagine overestimating allocatable leading to flapping.

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go (hunk 2 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/gce/gce_manager.go]: No, for debugging only. NodeInfo requires kube-proxy and daemonsets.

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go (hunk 1 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/gce/gce_manager.go]: Done.

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go (hunk 3 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/gce/gce_manager.go]: Done.

File: cluster-autoscaler/cloudprovider/gce/gce_manager_test.go (hunk 4 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/gce/gce_manager_test.go]: Done.

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go (hunk 5 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/gce/gce_manager.go]: Done.

**Comment by MaciekPytel** [comment]
lgtm

**Comment by fgrzadkowski** [comment]
@mwielgus I was under the impression that we didn't yet agree on the design. Did I misunderstood? Do we understand how allocatable is generated? Do we know which labels are not passed through `KUBE_ENV`?

**Comment by mwielgus** [comment]
@fgrzadkowski  we didn't agree yet on some of the details. This however doesn't mean that no code should be written. There are todos in the code that will be filled.

**Comment by mwielgus** [comment]
As of labels - yes - the code that populates all other is here:

https://github.com/kubernetes/kubernetes/blob/ceff8d8d4d7ac271cd03dcae73edde048a685df5/pkg/kubelet/kubelet_node_status.go#L196

**Comment by mwielgus** [comment]
As of allocatable - also the code was tracked down:
https://github.com/kubernetes/kubernetes/blob/c20e63bfb98fecef7461dbaf8ed52e31fe12cd11/pkg/kubelet/cm/node_container_manager.go#L184

However, for a moment we will live with allocatable = capacity. We will fix it before release.

**Comment by fgrzadkowski** [comment]
Thanks for checking the code! Let's sync on Monday and agree on the details
you've mentioned.

--
Filip

**Pull request merged**